---
title: "BCalm and analyze your MPRA data"
author: "Pia Keukeleire"
date: "`r format(Sys.time(), '%B %d, %Y')`"
package: "`r Githubpkg('kircherlab/BCalm')` (v0.1.0)"
bibliography: bcalm.bib
abstract: >
  A guide on how to use `BCalm` for analyzing massively parallel reporter assays (MPRA) data.
vignette: >
  %\VignetteIndexEntry{BCalm: A User's Guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE, crop=NULL)
```

```{r loading packages, echo=TRUE}
library(BCalm)
library(dplyr)
library(ggplot2)
```

# Introduction

The `r Githubpkg('kircherlab/BCalm')` package provides a framework for analyzing data from Massively Parallel Reporter Assays (MPRA) and is built on top of the mpra package. BCalm adapts the existing mpralm method but enhances it by modeling individual barcode counts rather than aggregating counts per sequence. Furthermore, the package includes a set of pre-processing functions and plotting capabilities, facilitating the visualization and interpretation of results. BCalm is more robust to outlier MPRA counts. Variant and element analysis are both shown below together with a significance test of elements against a control group (e.g. negative controls).

## Citing BCalm
The BCalm package is still unpublished, citing details will be provided later. When using BCalm, please cite the `mpra` package [@mpralm] and the limma-voom framework [@voom].

## Additional information for the installation
The package is currently available on GitHub and can be installed using remotes [@remotes] or devtools [@devtools].
The package requires R >= 3.5, <= 4.4.0.
If you have any trouble with the provided package feel free to let us know by creating an issue directly in the [BCalm GitHub repository](https://github.com/kircherlab/BCalm).
To display the vignette correctly, the `ggplot2` package is required.

# Preprocessing data

The first dataframe contains as small subset of a lentiMPRA dataset performed within HepG2 cells with three technical replicates (IGVF accession identifier: IGVFSM9009DVDG). Sequences tested in this experiment aim to capture variant effects across tens of thousands of candidate cis-regulatory element (cCRE) sequences of 200 base pair (bp) length.
The input files used here were obtained from  `r Githubpkg('kircherlab/MPRAsnakeflow')`, a Snakemake workflow produced as part of the Impact of Genomic Variation on Function (IGVF) Consortium. MPRAsnakeflow is a comprehensive pipeline which performs both the assignment of barcodes to the designed oligos and the preparation of count tables of DNA and RNA counts based on the observed number of barcodes within the targeted DNA and RNA sequencing (modified from [@Gordon2020]).

```{r First dataset}
data("BcSetExample")
nr_reps = 3
# show the data
print(head(BcSetExample))
```

In general, any sequence can be tested using an MPRA. Possible analyses can be differentiated by whether they compare the activity of different conditions of the same region, such as variant testing, or whether they compare the activities of different regions.
For element testing, the tested sequences in this vignette are compared to a group of negative control sequences (known to have low activity in HepG2) and the sequences to be tested.
We show the usage of BCalm on a variant dataset as well as an element dataset in this vignette. First, we show how to correctly preprocess the data.

## Variant testing

To prepare data for variant testing, we use the `create_var_df` function from BCalm. This function requires a mapping dataframe with information linking each reference allele to its corresponding alternative allele. Here, we use `MapExample`, a dataframe containing three essential columns: `ID`, `REF`, and `ALT`. This setup provides the necessary reference and alternative allele data to enable variant analysis.
```{r Variant Map}
# load the variant map
data("MapExample")
# show the data
print(head(MapExample))

# create the variant dataframe by adding the variant ID to the DNA and RNA counts
var_df <- create_var_df(BcSetExample, MapExample)
# show the data
print(head(var_df))
```

Optionally, downsampling can be performed to our dataframe `var_df` now. The function `downsample_barcodes` allows users to reduce the number of barcodes while retaining a representative subset. This way, the number of barcodes of oligos with many barcodes are reduced, which simplifies the data handling and reduces the sparseness of the data table (i.e. increased speed and reduced memory requirements). The degree of downsampling can be controlled by adjusting the sampling rate, which is expressed as a percentile value `percentile`, with a default of 0.975.
The `id_column_name` argument specifies the column in the input data frame that contains the unique identifiers for each variant (here `variant_id`).

```{r Variant Downsampling}
var_df <- downsample_barcodes(var_df, id_column_name="variant_id")
```

After downsampling the barcode counts in our dataset, we can prepare the data for analysis using the `create_dna_df` and `create_rna_df` functions.
Only six rows are shown here (original size of the dataframe 996 Ã— 474).
```{r Creating dna dataframe from var_df}
dna_var <- create_dna_df(var_df)
print(head(dna_var))
```
```{r Creating rna dataframe from var_df}
rna_var <- create_rna_df(var_df)
print(head(rna_var))
```
Now we create the MPRAset used as input to BCalm.

```{r Variant MPRASet}
# create the variant specific MPRAset
BcVariantMPRASetExample <- MPRASet(DNA = dna_var, RNA = rna_var, eid = row.names(dna_var), barcode = NULL)
```

## Element testing

The dataset is the same one we used above, but we have to add labels to the data to distinguish between control and test groups, thus allowing us to easily identify and compare these different groups in the analysis later.
```{r LabelsVec}
data(LabelExample)
table(LabelExample)
print(head(LabelExample))
```

Once again, we perform downsampling on this dataset using the `downsample_barcodes` function.
```{r Downsampling}
elem_df <- downsample_barcodes(BcSetExample)
```

As before, we use `create_dna_df` and `create_rna_df` to format the data correctly for the `MPRASet` function. However, this time we specify `id_column_name = "name"` since the default, `id_column_name = "variant_id"`, does not match our data format.
```{r Creating dataframes}
dna_elem <- create_dna_df(elem_df, id_column_name="name")
rna_elem <- create_rna_df(elem_df, id_column_name="name")
```

To compare between test and control, we need to add the labels to the MPRASet.

```{r Element MPRASet}
BcLabelMPRASetExample <- MPRASet(DNA = dna_elem, RNA = rna_elem, eid = row.names(dna_elem), barcode = NULL, label=LabelExample)
```

With the data prepared and preprocessed, we now have the foundation to conduct our analysis.

# Analysis
In this section we get to see the usage of the `mpralm` and the `fit_elements` functions. We take the `MPRASet` created in the preprocessing chapter. BCalm allows us to analyze individual barcode counts as separate samples, capturing additional data variation and potentially increasing statistical power.

## Variant Analysis
We will start with variant testing. In order to achieve this, we employ the `mpralm` function. Which column belongs to which replicate is described in a blocking vector, also used to normalize the counts per replicate.
The design matrix gives information which count comes from the reference and which from the alternative allele.

```{r Fit Variants}
bcs <- ncol(dna_var) / nr_reps
design <- data.frame(intcpt = 1, alt = grepl("alt", colnames(BcVariantMPRASetExample)))
block_vector <- rep(1:nr_reps, each=bcs)
mpralm_fit_var <- mpralm(object = BcVariantMPRASetExample, design = design, aggregate = "none", normalize = TRUE, model_type = "corr_groups", plot = FALSE, block = block_vector)

top_var <- topTable(mpralm_fit_var, coef = 2, number = Inf)
print(head(rna_var))
```

```{r Volcano Plot}
ggplot(top_var, aes(x = logFC, y = -log10(P.Value))) +
  geom_point(alpha = 0.6)
```

## Element Analysis

BCalm provides the function `fit_elements`. It takes the `MPRASet` object as input and applies the statistical modeling. Again the `block_vector` gives reference which barcode belongs to which replicate.
We again set `normalize = TRUE` to perform total count normalization on the RNA and DNA libraries.

```{r Fit elements, fig.width=8, fig.height=4}
bcs <- ncol(dna_elem) / nr_reps
block_vector <- rep(1:nr_reps, each=bcs)
mpralm_fit_elem <- fit_elements(object = BcLabelMPRASetExample, normalize=TRUE, block = block_vector, plot = FALSE)
```

### Visualisation and results
In this section, we will examine the visualization of our analysis results using the `mpra_treat` and `plot_groups` functions.
To visualize our results, we utilize the `plot_groups` function, which allows us to compare logratios for each group. We use the results from `fit_elements` above. As negative controls we use `"control"` and as test `"tested"`.

```{r Visualization, fig=TRUE, fig.width=8, fig.height=4, warning=FALSE}
plot_groups(mpralm_fit_elem, 0.975, neg_label="control", test_label="tested")
```

The `mpra_treat()` function reimplements the `treat()` function from the limma package. This function performs a t-test with a specified threshold, making it especially useful for identifying elements with significant differential activity in MPRA data.
```{r MPRA treat}
treat <- mpra_treat(mpralm_fit_elem, 0.975, neg_label="control")
result <- topTreat(treat, coef = 1, number = Inf)
head(result)
```

# Session Info

```{r sessionInfo, results='asis', echo=FALSE}
sessionInfo()
```

# References
